File: SAMTextView/SAMTextView.m
//
//  SAMTextView.m
//  SAMTextView
//
//  Created by Sam Soffes on 8/18/10.
//  Copyright 2010-2013 Sam Soffes. All rights reserved.
//

#import "SAMTextView.h"

@implementation SAMTextView

#pragma mark - Accessors

@synthesize placeholder = _placeholder;
@synthesize placeholderTextColor = _placeholderTextColor;

- (void)setText:(NSString *)string {
	[super setText:string];
	[self setNeedsDisplay];
}


- (void)insertText:(NSString *)string {
	[super insertText:string];
	[self setNeedsDisplay];
}


- (void)setAttributedText:(NSAttributedString *)attributedText {
	[super setAttributedText:attributedText];
	[self setNeedsDisplay];
}


- (void)setPlaceholder:(NSString *)string {
	if ([string isEqual:_placeholder]) {
		return;
	}

	_placeholder = string;
	[self setNeedsDisplay];
}


- (void)setContentInset:(UIEdgeInsets)contentInset {
	[super setContentInset:contentInset];
	[self setNeedsDisplay];
}


- (void)setFont:(UIFont *)font {
	[super setFont:font];
	[self setNeedsDisplay];
}


- (void)setTextAlignment:(NSTextAlignment)textAlignment {
	[super setTextAlignment:textAlignment];
	[self setNeedsDisplay];
}


#pragma mark - NSObject

- (void)dealloc {
	[[NSNotificationCenter defaultCenter] removeObserver:self name:UITextViewTextDidChangeNotification object:self];
}


#pragma mark - UIView

- (id)initWithCoder:(NSCoder *)aDecoder {
	if ((self = [super initWithCoder:aDecoder])) {
		[self initialize];
	}
	return self;
}


- (id)initWithFrame:(CGRect)frame {
	if ((self = [super initWithFrame:frame])) {
		[self initialize];
	}
	return self;
}


- (void)drawRect:(CGRect)rect {
	[super drawRect:rect];

	if (self.text.length == 0 && self.placeholder) {
		rect = [self placeholderRectForBounds:self.bounds];

		UIFont *font = self.font ? self.font : self.typingAttributes[NSFontAttributeName];

		// Draw the text
		[self.placeholderTextColor set];
        if ([_placeholder respondsToSelector:@selector(drawInRect:withAttributes:)]){
            NSMutableParagraphStyle *paragraphStyle = [[NSMutableParagraphStyle alloc] init];
            paragraphStyle.lineBreakMode = NSLineBreakByTruncatingTail;
            paragraphStyle.alignment = self.textAlignment;
            NSDictionary *attributes = @{NSFontAttributeName : self.font,
                                         NSParagraphStyleAttributeName : paragraphStyle};
            [_placeholder drawInRect:rect withAttributes:attributes];
        }
        else {
            [self.placeholder drawInRect:rect withFont:font lineBreakMode:NSLineBreakByTruncatingTail alignment:self.textAlignment];
        }
	}
}


#pragma mark - Placeholder

- (CGRect)placeholderRectForBounds:(CGRect)bounds {
	// Inset the rect
	CGRect rect = UIEdgeInsetsInsetRect(bounds, self.contentInset);

	UIFont *font = self.font ? self.font : self.typingAttributes[NSFontAttributeName];

	if (self.typingAttributes) {
		NSParagraphStyle *style = self.typingAttributes[NSParagraphStyleAttributeName];
		if (style) {
			rect.origin.x += style.headIndent;
			rect.origin.y += style.firstLineHeadIndent;
		}
	}

	return rect;
}


#pragma mark - Private

- (void)initialize {
	[[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(textChanged:) name:UITextViewTextDidChangeNotification object:self];

	self.placeholderTextColor = [UIColor colorWithWhite:0.702f alpha:1.0f];
}


- (void)textChanged:(NSNotification *)notification {
	[self setNeedsDisplay];
}

@end

File: SAMTextView/SAMTextView.swift
//
//  SAMTextView.swift
//  SAMTextView
//
//  Created by Sam Soffes on 8/18/10.
//  Copyright 2010-2014 Sam Soffes. All rights reserved.
//

import UIKit

/**
 UITextView subclass that adds placeholder support like UITextField has.
 */
public class SAMTextView : UITextView {
    public required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        initialize()
    }
    
    override init(frame: CGRect, textContainer: NSTextContainer?) {
        super.init(frame: frame, textContainer: textContainer)
        initialize()
    }
    
    /**
     The string that is displayed when there is no other text in the text view. This property reads and writes the
     attributed variant.
     
     The default value is `nil`.
     */
    public var placeholder: String? {
        get {
            return attributedPlaceholder?.string
        }
        set(newPlaceholder) {
            guard let newPlaceholder = newPlaceholder else { return }
            guard newPlaceholder != attributedPlaceholder?.string else { return }
        
            var attributes = [String : AnyObject]()
            if isFirstResponder() {
                attributes = typingAttributes
            } else {
                attributes[NSFontAttributeName] = font
                attributes[NSForegroundColorAttributeName] = UIColor(white: 0.702, alpha: 1)
            
                if textAlignment != NSTextAlignment.Left {
                    let paragraph = NSMutableParagraphStyle()
                    paragraph.alignment = textAlignment
                    attributes[NSParagraphStyleAttributeName] = paragraph
                }
            }
            attributedPlaceholder = NSAttributedString(string: newPlaceholder, attributes: attributes)
        }
    }
    
    /**
     The attributed string that is displayed when there is no other text in the text view.
     
     The default value is `nil`.
     */
    var attributedPlaceholder: NSAttributedString? {
        didSet {
            setNeedsDisplay()
        }
    }
    
    public override var text: String! {
        didSet {
            setNeedsDisplay()
        }
    }
    
    public override var attributedText: NSAttributedString! {
        didSet {
            setNeedsDisplay()
        }
    }
    
    public override var contentInset: UIEdgeInsets {
        didSet {
            setNeedsDisplay()
        }
    }
    
    public override var font: UIFont? {
        didSet {
            setNeedsDisplay()
        }
    }
    
    public override var textAlignment: NSTextAlignment {
        didSet {
            setNeedsDisplay()
        }
    }
    
    public override func insertText(text: String) {
        super.insertText(text)
        setNeedsDisplay()
    }
    
    // MARK: - Deinit
    
    deinit {
        NSNotificationCenter.defaultCenter().removeObserver(self, name: UITextViewTextDidChangeNotification, object: self)
    }
    
    // MARK: - UIView
    
    public override func drawRect(rect: CGRect) {
        super.drawRect(rect)
        
        // Draw placeholder if necessary
        if text.characters.count == 0 && attributedPlaceholder != nil {
            let placeholderRect = placeholderRectForBounds(bounds)
            attributedPlaceholder?.drawInRect(placeholderRect)
        }
    }
    
    public override func layoutSubviews() {
        super.layoutSubviews()
        
        // Redraw placeholder text when the layout changes if necessary
        if attributedPlaceholder != nil && text.characters.count == 0 {
            setNeedsDisplay()
        }
    }
    
    // MARK: - Private
    
    private func textChanged() {
        setNeedsDisplay()
    }
    
    private func initialize() {
        NSNotificationCenter.defaultCenter().addObserverForName(UITextViewTextDidChangeNotification, object: self, queue: nil) { [unowned self] notification in
            self.textChanged()
        }
    }
    
    /**
     Returns the drawing rectangle for the text views’s placeholder text.
     
     - parameter bounds: The bounding rectangle of the receiver.
     
     - returns: The computed drawing rectangle for the placeholder text.
     */
    private func placeholderRectForBounds(bounds: CGRect) -> CGRect {
        var rect = UIEdgeInsetsInsetRect(bounds, contentInset)
        
        if respondsToSelector("textContainer") {
            rect = UIEdgeInsetsInsetRect(rect, textContainerInset)
            let padding = textContainer.lineFragmentPadding
            rect.origin.x += padding
            rect.size.width -= padding * 2.0
        } else {
            if contentInset.left == 0.0 {
                rect.origin.x += 8.0
            }
            rect.origin.y += 8.0
        }
        
        return rect
    }
}
